File: ProblemSolvingKotlin/src/Module1.kt
import java.math.BigInteger

/**
 * Created by keyurgolani on 6/1/17.
 */

fun sumNumbers(values:List<Int>): Int {
    var answer: Int = 0
    for (value in values) {
        answer += value
    }
    return answer
}

fun maxPairwiseProduct(values: List<Int>): Int {
    var maxOne = Int.MIN_VALUE
    var maxTwo = Int.MIN_VALUE
    var minOne = Int.MAX_VALUE
    var minTwo = Int.MAX_VALUE
    for (value in values) {
        if (value >= maxOne) {
            maxTwo = maxOne
            maxOne = value
        } else if (value >= maxTwo) {
            maxTwo = value
        }
        if (value <= minOne) {
            minTwo = minOne
            minOne = value
        } else if (value <= minTwo) {
            minTwo = value
        }
    }
    val maxProduct = maxOne * maxTwo
    val minProduct = minOne * minTwo
    return if (minProduct > maxProduct) minProduct else maxProduct
}

fun maxPairwiseSum(values: List<Int>): Int {
    var maxOne = Int.MIN_VALUE
    var maxTwo = Int.MIN_VALUE
    for (value in values) {
        if (value >= maxOne) {
            maxTwo = maxOne
            maxOne = value
        } else if (value >= maxTwo) {
            maxTwo = value
        }
    }
    return maxOne + maxTwo
}

fun fibonacciSeriesTill(position: Int): List<BigInteger> {
    val fibonacci: MutableList<BigInteger> = mutableListOf(BigInteger.ZERO, BigInteger.ONE)
    if (position > 1) {
        (2..position - 1).forEach { fibonacci.add(it, fibonacci[it - 1] + fibonacci[it - 2]) }
    }
    return fibonacci
}

fun fibonacciNumberAt(position: Int): BigInteger {
    val fibonacci: MutableList<BigInteger> = mutableListOf(BigInteger.ZERO, BigInteger.ONE)
    if (position > 1) {
        (2..position).forEach {
            fibonacci[1] = fibonacci[0] + fibonacci[1]
            fibonacci[0] = fibonacci[1] - fibonacci[0]
        }
    }
    return fibonacci[1]
}

fun lastDigitOfFibonacciNumberAt(position: Int): Int {
    val fibonacci: MutableList<Int> = mutableListOf(0, 1)
    if (position > 1) {
        (2..position - 1).forEach {
            fibonacci[1] = fibonacci[0] % 10 + fibonacci[1] % 10
            fibonacci[0] = fibonacci[1] % 10 - fibonacci[0] % 10
        }
    }
    return (((fibonacci[1] % 10) + 10) % 10)
}

fun fibonacciNumberAtModulo(position: BigInteger, number: Int): BigInteger {
    if (number == 0) {
        return fibonacciNumberAt(position.toInt())
    } else if (number == 1) {
        return BigInteger.ZERO
    } else {
        val s = mutableListOf<BigInteger>(BigInteger.ZERO, BigInteger.ONE)
        var idx: Int = 0
        while (!(idx > 0 && s[idx] == BigInteger.ZERO && s[idx + 1] == BigInteger.ONE)) {
            s.add(((s[idx] % number.toBigInteger() + s[idx + 1] % number.toBigInteger()) % number.toBigInteger()))
            idx++
        }
        return s[(position % idx.toBigInteger()).toInt()]
    }
}

fun lastDigitOfSumOfFibonacciNumbersTill(limit: Int): Int {
    var sum:BigInteger = BigInteger.ZERO
    val fibonacci = mutableListOf<BigInteger>(BigInteger.ZERO, BigInteger.ONE)
    (0..limit).forEach {
        if (it > 1) {
            fibonacci[1] = fibonacci[0] + fibonacci[1]
            fibonacci[0] = fibonacci[1] - fibonacci[0]
            sum += fibonacci[1] % BigInteger.TEN
        } else {
            sum += fibonacci[it]
        }
    }
    return (sum % BigInteger.TEN).toInt()
}

fun lastDigitOfPartialSumOfFibonacciNumbers(start: Int, end: Int): Int {
    var sum:BigInteger = BigInteger.ZERO
    val fibonacci = mutableListOf<BigInteger>(BigInteger.ZERO, BigInteger.ONE)
    (0..start - 1).forEach {
        if (it > 1) {
            fibonacci[1] = fibonacci[0] + fibonacci[1]
            fibonacci[0] = fibonacci[1] - fibonacci[0]
        }
    }
    (start..end).forEach {
        if (it > 1) {
            fibonacci[1] = fibonacci[0] + fibonacci[1]
            fibonacci[0] = fibonacci[1] - fibonacci[0]
            sum += fibonacci[1] % BigInteger.TEN
        } else {
            sum += fibonacci[it]
        }
    }
    return (sum % BigInteger.TEN).toInt()
}

fun gcdOf(number1: BigInteger, number2: BigInteger): BigInteger {
    var num1 = number1
    var num2 = number2
    while (num2 != BigInteger.ZERO) {
        val temp = num2
        num2 = num1 % num2
        num1 = temp
    }
    return num1
}

fun lcmOf(number1: BigInteger, number2: BigInteger): BigInteger {
    return number1 * number2 / gcdOf(number1, number2)
}

File: ProblemSolvingPython/Module1.py
import sys
import Utilities as Util


def sum_of_numbers(values):
    """
    Function that accepts an array of integer values and returns the sum of all the integer values
    :param values: an array of integer values to be summed
    :return: sum of all integer values passed as input parameter
    """
    answer = 0
    for value in values:
        answer += value
    return answer


def max_pairwise_product(values):
    """
    Function that accepts an array of integer values and returns maximum pairwise product from the array
    :param values: an array of integer values
    :return: maximum pairwise product from the input values
    """
    if len(values) <= 1:
        return None
    max_one = -sys.maxint - 1
    max_two = -sys.maxint - 1
    min_one = sys.maxint
    min_two = sys.maxint
    for value in values:
        if value >= max_one:
            max_two = max_one
            max_one = value
        if max_one > value >= max_two:
            max_two = value
        if value <= min_one:
            min_two = min_one
            min_one = value
        if min_one < value <= min_two:
            min_two = value

    print max_one, max_two, min_one, min_two
    min_product = min_one * min_two
    max_product = max_one * max_two
    return min_product if min_product > max_product else max_product


def max_pairwise_sum(values):
    """
    Function that accepts an array of integer values and returns maximum pairwise sum from the array
    :param values: an array of integer values
    :return: maximum pairwise sum from the input values
    """
    if len(values) <= 1:
        return None
    max_one = -sys.maxint - 1
    max_two = -sys.maxint - 1
    for value in values:
        if value >= max_one:
            max_two = max_one
            max_one = value
    return max_one + max_two


def fibonacci_series(limit):
    """
    Yields the value of fibonacci series at current index below the limit
    :param limit: limit integer to generate fibonacci series upto
    :return: yields fibonacci value of current idx
    """
    fibonacci = [0, 1]
    for idx in range(limit+1):
        if idx <= 1:
            yield fibonacci[idx]
        else:
            fibonacci.append(fibonacci[idx - 1] + fibonacci[idx - 2])
            yield fibonacci[idx]


def fibonacci_number(position):
    """
    Returns the value of number at given position in fibonacci series
    :param position: position to return number at from fibonacci series
    :return: returns the integer number at position from fibonacci series
    """
    fibonacci = [0, 1]
    for idx in range(position+1):
        if idx > 1:
            fibonacci.append(fibonacci[idx - 1] + fibonacci[idx - 2])
    return fibonacci[position]


def fibonacci_number_last_digit(position):
    """
    Returns the last digit of number at given position in fibonacci series
    :param position: position to return number at from fibonacci series
    :return: returns the integer last digit of number at position from fibonacci series
    """
    fibonacci = [0, 1]
    for idx in range(position+1):
        if idx > 1:
            fibonacci.append((fibonacci[idx - 1] + fibonacci[idx - 2]) % 10)
    return fibonacci[position]


def fibonacci_number_modulo(position, number):
    """
    Returns the fibonacci number at given position modulo given number
    :param position: position for the fibonacci number in fibonacci series
    :param number: number to modulo fibonacci number with
    :return: integer modulo value of fibonacci number at position with given number
    """
    if number == 0:
        return fibonacci_number(position)
    elif number == 1:
        return 0
    else:
        pattern_position = position % Util.get_pisano_period(number)
        return fibonacci_number(pattern_position) % number


def last_digit_of_sum_of_fibonacci_numbers_till(limit):
    """
    Returns last digit of sum of fibonacci numbers till limit
    :param limit: limit till which to sum the fibonacci numbers
    :return: returns integer last digit of sum of all fibonacci numbers till given limit
    """
    sum = 0
    fibonacci = [0, 1]
    for idx in range(limit + 1):
        if idx <= 1:
            sum += fibonacci[idx] % 10
        else:
            fibonacci[0], fibonacci[1] = fibonacci[1], fibonacci[0] + fibonacci[1]
            sum += fibonacci[1] % 10
    return sum % 10


def last_digit_of_partial_sum_of_fibonacci_numbers(start, end):
    """
    Returns last digit of partial sum of fibonacci numbers from start till end
    :param start: start position of the partial sum in series
    :param end: end position of the partial sum in series
    :return: returns integer last digit of partial sum of all fibonacci numbers from start till end
    """
    sum = 0
    fibonacci = [0, 1]
    for idx in range(start):
        if idx > 1:
            fibonacci[0], fibonacci[1] = fibonacci[1], fibonacci[0] + fibonacci[1]
    for idx in range(start, end+1):
        if idx <= 1:
            sum += fibonacci[idx] % 10
        else:
            fibonacci[0], fibonacci[1] = fibonacci[1], fibonacci[0] + fibonacci[1]
            sum += fibonacci[1] % 10
    return sum % 10



def gcd(number_one, number_two):
    """
    Gives Greatest Common Divisor for the numbers passed
    Used Euclidean Algorithm
    :param number_one: Integer Number one for GCD
    :param number_two: Integer Number two for GCD
    :return: Integer Greatest Common Divisor for two numbers input
    """
    while number_two != 0:
        number_one, number_two = number_two, number_one % number_two
    return number_one


def lcm(number_one, number_two):
    """
    Gives Greatest Common Divisor for the numbers passed
    Used Euclidean Algorithm
    :param number_one: Integer Number one for GCD
    :param number_two: Integer Number two for GCD
    :return: Integer Greatest Common Divisor for two numbers input
    """
    return number_one * number_two / gcd(number_one, number_two)
